Generalise and clean up blueprint customization validation for image types (HMS-6871)

achilleas-k opened the pull request:
Year-old branch that I never finished.  I remembered it today during a discussion with @supakeen, @mvo5, and @schuellerf.  Opening as draft because I don't want it to disappear or forget about it again.

lzap commented on pkg/blueprint/customizations.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"reflect"
 	"strings"
+
+	"golang.org/x/exp/slices"
```
Part of Go 1.21, at least the used `Reverse` [function](https://pkg.go.dev/slices#Reverse).

achilleas-k replied:
As the title mentions, this is a year-old branch :laughing: 
But thanks, I'll try to get back to this soon and rebase.  It's going to need a lot of work.

mvo5 commented:
We will need to keep this open for reference as it will become relevant for the YAMLification and move to a "generic" distro everywhere.

lzap commented:
For the record, I have a ticket: https://issues.redhat.com/browse/HMS-8676

achilleas-k commented:
The validation functionality is ready now.  I need to finish this now by adding the customization lists to the image types.

lzap commented:
It looks good, tho, I am wondering why this patch was mentioned as a blocker for UBP. Or have I [commented](https://github.com/osbuild/images/pull/1216#issuecomment-3000024677) on a wrong PR?

achilleas-k commented:
> I am wondering why this patch was mentioned as a blocker for UBP

Is it?  I think we talked about it being relevant, but not a blocker.

bcl reviewed (changes requested):
I think this is pretty nice. customizations.partitioning_mode and customizations.sshkey are missing -- but they are not supported by the images blueprint so that's a separate problem.

lzap commented on pkg/distro/defs/fedora/distro.yaml:
```diff
@@ -1884,46 +1852,69 @@ image_types:
-            - "shadow-utils"  # includes passwd
+            - "xfsprogs"
+            - "xz"
           conditions:
```
Not sure how this hunk is related to the PR title, but okay.

lzap replied:
No hard feelings and I usually do not care as long as commits are separate, I was just curious. Feel free to keep it in!

achilleas-k commented on pkg/distro/validator.go:
```diff
@@ -0,0 +1,183 @@
+	return nil
+}
+
+func ValidateConfig(t ImageType, bp blueprint.Blueprint, _ ImageOptions) error {
```
I completely forgot about this comment.  Perhaps we can simplify in a followup.  It'd be nice to do a wider breakdown of our larger interfaces in general, right?

achilleas-k commented on pkg/blueprint/blueprint.go:
```diff
@@ -96,3 +102,15 @@ func (p Package) ToNameVersion() string {
+
+type BlueprintError struct {
+	// Reverse path to the customization that caused the error.
+	RevPath []string
```
A dedicated type for the RevPath sounds a bit too "overengineery" to me.  But perhaps making it a private field and adding an `Append()` (or `AddPath()`?) method to `BlueprintError` might be nice.

achilleas-k commented:
Responding to [this comment thread](https://github.com/osbuild/images/pull/1216#discussion_r2255439925) here for better visibility.

> All of these need to also have `name`, `version`, and `description` added, otherwise the most basic blueprint is rejected.

I added the support in a separate commit (I can squash it with the others if people feel it's cleaner) but, more importantly, I added descriptions to all test blueprints under `test/configs/`.  The description explains the purpose of each test configuration, which I think is a useful thing to have.

mvo5 commented on data/distrodefs/fedora/imagetypes.yaml:
```diff
@@ -532,6 +532,91 @@
+    # common options supported by all disk image types this includes everything
+    # that is not specific to installers or ostree-based images
+    supported_options_disk: &supported_options_disk
+      - "name"
```
(wondering) Should we make name/version/description implicit? They are not really concepts that are relevant for the image types, i.e. it would be weird to have an image type that refuses a blueprint with a "name"?

Similar wondering about "distro", it goes away in UBP and its a bit of a odd one, so maybe the image types should not list it?

bcl replied:
I think that hiding it makes it slightly more difficult to get up to speed with it. It's supposed to be a list of blueprint fields, just make it 1:1 without any magic.

mvo5 commented on pkg/distro/validator_test.go:
```diff
@@ -0,0 +1,1196 @@
+	"github.com/stretchr/testify/assert"
+)
+
+type TestImageType struct {
```
(nitpick) I would love if we had only one TestImageType, ideally from `test_distro` and with exported fields so that each test that needs an ImageType can just import and tweak it. But I understand its probably a bit cumbersome right now with the existing `test_distro`

achilleas-k replied:
Can we move this to a follow-up?  We could also solve this by breaking down the image type interface, which I would like to do at some point anyway.

mvo5 commented on test/configs/all-customizations.json:
```diff
@@ -1,6 +1,9 @@
   "name": "all-customizations",
   "blueprint": {
+    "name": "all-customizations",
+    "description": "Blueprint that uses (almost) all available customizations for the disk image types",
```
❤️

mvo5 reviewed (commented):
I really like how clean this is and how straightfoward the yaml is. I did not yet do a full in-depth review but have a few questions/ideas inline from my first read.

bcl reviewed (approved):
Looks good!

thozza commented:
Out of curiosity, why didn't you consider defining the supported customizations the other way around? IOW for the image type to define the customizations that it does not support? Do you think that it would make sense to also provide this approach in the future?

bcl commented:
> @bcl What would the supported options list look like for the net installer?

Right now the only things supported are FIPS and language.

mvo5 commented on pkg/distro/validator.go:
```diff
@@ -106,12 +106,139 @@ func validateSupportedConfig(supported []string, conf reflect.Value) *validation
+
+	requiredMap := make(map[string]bool)
+	subMap := make(map[string][]string)
+	for _, key := range required {
```
Yeah, there is no obvious way (I was thinking an iterator and callback and the callback would be specific to supported/required) - but then I'm absolutely overthinking this (sorry for that)

mvo5 reviewed (dismissed):
Thank you!

mvo5 commented on pkg/distro/validator.go:
```diff
@@ -106,12 +106,139 @@ func validateSupportedConfig(supported []string, conf reflect.Value) *validation
+			// missing value.
+			if value.IsZero() {
+				return &validationError{message: "required", revPath: []string{key}}
+			}
```
(note for other reviewers, this got also added)

bcl commented on data/distrodefs/fedora/imagetypes.yaml:
```diff
@@ -752,6 +752,91 @@
+      - "customizations.openscap"
+      - "customizations.repositories"
+      - "customizations.rpm"
+      - "customizations.services"
```
This is missing `customizations.sshkey` and `customizations.partition_mode` which I stumbled into by accident when trying a blueprint I had laying around from something else, not through any thorough examination so there may be others :/

supakeen replied:
The former only exists in `osbuild-composer` and (AFAIK) gets translated to a `customizations.user` we mention it as [deprecated](https://osbuild.org/docs/user-guide/blueprint-reference/#ssh-keys) in the documentation though I've gotten bitten by it before as well.

The latter is indeed missing and [documented to exist](https://osbuild.org/docs/user-guide/blueprint-reference/#partitioning-mode).

bcl reviewed (changes requested):
I wish there was an automated way to be sure all the correct options are supported. If we miss one then people's existing blueprints will break.

achilleas-k commented:
Updated:
- Rebased.
- Fixed conflict.
- Added `customizations.sshkey` to all supported lists that also include `customizations.user`.
- Added `customizations.partitioning_mode` to all supported lists that also include `customizations.filesystem`.

supakeen reviewed (approved):
Thank you.

mvo5 commented:
Its nice, having this will help with https://github.com/osbuild/images/pull/1786 - i.e. we can selectively allow kernel options for the installer types that currently do not support them even if the underlying code is all unified.
